unit/analyses/ai/ai.cpp

```diff
@@ -0,0 +1,310 @@
+/*******************************************************************\
+
+ Module: Unit tests for selectively retaining particular domains in ait
+
+ Author: Diffblue Ltd.
+
+\*******************************************************************/
+
+/// \file
+/// Unit tests for selectively retaining particular domains in ait
+
+#include <testing-utils/catch.hpp>
+
+#include <analyses/ai.h>
+
+#include <ansi-c/ansi_c_language.h>
+
+#include <goto-programs/goto_convert_functions.h>
+
+#include <langapi/mode.h>
+
+#include <util/arith_tools.h>
+#include <util/config.h>
+#include <util/c_types.h>
+
+/// A very simple analysis that counts executed instructions along a particular
+/// path, taking the max at merge points and saturating at 100 instructions.
+/// It should indicate that instructions not within a loop have a certain path
+/// length, and those reachable from a loop have a path length of 100
+/// (i.e. potentially infinity)
+class instruction_counter_domaint : public ai_domain_baset
+{
+public:
+
+  std::unique_ptr<int> path_length;
```

Comment:
`optional`?

Reply:
Can `path_length` be negative?

Reply:
Also; I'm probably being thick, but why a pointer, why not just an int?